File: libraries/go/manifest.go
package vsync

import "fmt"

// RQEM0001 manifest pointer-seal — read path only (v0.2 §3, v0.4):
//
//	bytes 0..7    magic "RQEM0001"
//	bytes 8..22   15-char ASCII timestamp "YYYYMMDD-HHmmss"
//	bytes 23..N   payload (opaque)
//
// VerifyManifestAgainstRemoteTS is the load-bearing anti-rollback check —
// an attacker with bucket-write but no key can't quietly swing the
// manifest pointer at a renamed older bundle because the embedded ts won't
// match the URL/key the lib fetched it from.

var rqem0001Magic = []byte{'R', 'Q', 'E', 'M', '0', '0', '0', '1'}

const (
	rqem0001MagicLen  = 8
	rqem0001TSLen     = 15
	rqem0001HeaderLen = rqem0001MagicLen + rqem0001TSLen // 23
)

// UnwrapRQEM0001 parses the envelope and returns (ts, payload). Does NOT
// verify against a remote ts — see VerifyManifestAgainstRemoteTS for that.
func UnwrapRQEM0001(blob []byte) (string, []byte, error) {
	if len(blob) < rqem0001HeaderLen {
		return "", nil, fmt.Errorf("%w: RQEM0001 manifest too short (%d bytes < %d)",
			ErrBundleCorrupt, len(blob), rqem0001HeaderLen)
	}
	if !equalBytes(blob[:rqem0001MagicLen], rqem0001Magic) {
		return "", nil, fmt.Errorf("%w: RQEM0001 magic prefix mismatch", ErrBundleCorrupt)
	}
	tsBytes := blob[rqem0001MagicLen:rqem0001HeaderLen]
	for _, c := range tsBytes {
		if c > 0x7f {
			return "", nil, fmt.Errorf("%w: RQEM0001 timestamp is not ASCII", ErrBundleCorrupt)
		}
	}
	ts := string(tsBytes)
	payload := blob[rqem0001HeaderLen:]
	return ts, payload, nil
}

// VerifyManifestAgainstRemoteTS unwraps and confirms the embedded ts matches
// remoteTS — i.e., the bucket-side object key the manifest was fetched from.
func VerifyManifestAgainstRemoteTS(blob []byte, remoteTS string) (string, []byte, error) {
	ts, payload, err := UnwrapRQEM0001(blob)
	if err != nil {
		return "", nil, err
	}
	if ts != remoteTS {
		return "", nil, fmt.Errorf("%w: RQEM0001 embedded ts %q != remote ts %q — possible rollback or torn write",
			ErrBundleCorrupt, ts, remoteTS)
	}
	return ts, payload, nil
}

// wrapManifestForTest mints a manifest for the unit suite. Production
// callers never write manifests — the CLI is the canonical writer.
func wrapManifestForTest(ts string, payload []byte) []byte {
	out := make([]byte, 0, rqem0001HeaderLen+len(payload))
	out = append(out, rqem0001Magic...)
	tsBytes := []byte(ts)
	if len(tsBytes) != rqem0001TSLen {
		// Pad / truncate so tests that pass weird strings still produce
		// well-framed output and the unwrap path exercises real bytes.
		buf := make([]byte, rqem0001TSLen)
		copy(buf, tsBytes)
		tsBytes = buf
	}
	out = append(out, tsBytes...)
	out = append(out, payload...)
	return out
}

File: libraries/typescript/src/manifest.ts
// RQEM0001 manifest pointer-seal — read path only (v0.2 §3 / v0.12).
//
//   bytes 0..7    magic "RQEM0001"
//   bytes 8..22   15-char ASCII timestamp "YYYYMMDD-HHmmss"
//   bytes 23..N   payload (opaque)
//
// `verifyAgainstRemoteTs` is the load-bearing anti-rollback check — an
// attacker with bucket-write but no decryption key cannot swing the
// `<prefix>manifest` pointer at a renamed older bundle without the
// embedded ts disagreeing with the remote ts.

import { BundleCorruptError } from "./errors.js";

const MAGIC = Buffer.from("RQEM0001", "ascii");
const TS_LEN = 15;
const HEADER_LEN = MAGIC.length + TS_LEN; // 23

export type ManifestParts = {
  ts: string;
  payload: Uint8Array;
};

/**
 * Parse the RQEM0001 envelope; return {ts, payload}.
 * Throws BundleCorruptError on missing magic or truncated input.
 * Does NOT verify against a remote ts — caller does that via
 * verifyAgainstRemoteTs.
 */
export function unwrapRqem0001(blob: Uint8Array): ManifestParts {
  if (blob.byteLength < HEADER_LEN) {
    throw new BundleCorruptError(
      `RQEM0001 manifest too short: ${blob.byteLength} bytes (need at least ${HEADER_LEN})`,
    );
  }
  const buf = Buffer.from(blob.buffer, blob.byteOffset, blob.byteLength);
  if (!buf.subarray(0, MAGIC.length).equals(MAGIC)) {
    throw new BundleCorruptError(
      "RQEM0001 manifest: magic prefix mismatch — not a vsync manifest",
    );
  }
  const ts = buf.subarray(MAGIC.length, HEADER_LEN).toString("ascii");
  const payload = new Uint8Array(
    buf.subarray(HEADER_LEN).buffer,
    buf.subarray(HEADER_LEN).byteOffset,
    buf.subarray(HEADER_LEN).byteLength,
  );
  return { ts, payload };
}

/**
 * Unwrap + verify the embedded ts equals `remoteTs`. A pointer-rollback
 * attack will fail this check.
 */
export function verifyAgainstRemoteTs(
  blob: Uint8Array,
  remoteTs: string,
): ManifestParts {
  const { ts, payload } = unwrapRqem0001(blob);
  if (ts !== remoteTs) {
    throw new BundleCorruptError(
      `RQEM0001 manifest: embedded ts ${JSON.stringify(ts)} != remote ts ${JSON.stringify(remoteTs)} — possible pointer-rollback attack or torn bucket write`,
    );
  }
  return { ts, payload };
}
